### PR TITLE
[Performance] Avoid reading all mailboxes upon GetMessageList

### PR DESCRIPTION
**Pull request by chibenwa**
When the user specify a single mailbox inMailboxes filter, we still read
all mailboxes, resulting in many unwanted queries

### CODE REVIEW

**Comment by chibenwa** [comment]
```
[01d8482ff001a52273bd7166ead1211c205ca643] ADD failed: stat /var/lib/docker/tmp/docker-builder542699651/destination/james-server-app-3.5.0-SNAPSHOT-app.zip: no such file or directory
```

:green_apple: if rebased

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java [hunk 1]
```diff
@@ -687,13 +687,29 @@ private boolean hasChildIn(Mailbox parentMailbox, List<Mailbox> mailboxesWithPat
     }
 
     private Stream<MailboxId> getInMailboxes(ImmutableSet<MailboxId> inMailboxes, MailboxSession session) throws MailboxException {
-       if (inMailboxes.isEmpty()) {
+        if (inMailboxes.isEmpty()) {
             return getAllReadableMailbox(session);
         } else {
+            if (inMailboxes.size() == 1) {
+                MailboxId mailboxId = inMailboxes.stream().findAny().get();
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java:694]: `Iterables.getOnlyElement`?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java [hunk 2]
```diff
@@ -687,13 +687,29 @@ private boolean hasChildIn(Mailbox parentMailbox, List<Mailbox> mailboxesWithPat
...
+                return filterReadable(mailboxId, session);
+            }
             return getAllReadableMailbox(session).filter(inMailboxes::contains);
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java:697]: Why don't we push the predicate down to allow the search to use it?
The optimization you did assumes things that only the implementation should know (like latency on `hasRight` vs latency on `searchMailboxes` for example)

Comment (chibenwa) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java:697]: The Mapper api expects a condition on the name, not on the ids.

I'm unsure of the result you expect.

**Comment by chibenwa** [comment]
(integration test) :green_apple:

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java (hunk 2 above)
Comment (mbaechler) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java:697]: This is a typical case when talking about search: you have a predicate (or a clause) for search and the backend don't know how to implement it.
Either you implement it un-efficiently at the top level and backends won't be able to optimize for that case: quick to implement but will need more work in the future.
Or you implement a method for that in the backends that actually do the same dumb filter but you open the door to an efficient implementation in backends.

I suggest we do the second one.

To talk about code, as you know, I'm not a big fan of `Mapper` concept.

We already have this nice search interface we need to push down the predicates, it's the `search` method.

Let's just create 3 protected methods:
* listAllMailboxes
* searchSingleMailbox
* searchMailboxesById

It allows any type extending StoreMailboxManager to optimize  some cases.

**Comment by chibenwa** [comment]
```
[db01eb7e7cf2a5f03f82bf6779ea38becf953bb0] [ERROR] src/main/java/org/apache/james/mailbox/store/mail/MailboxMapper.java:[99] (whitespace) EmptyLineSeparator: ';' should be separated from previous statement.
[db01eb7e7cf2a5f03f82bf6779ea38becf953bb0] [ERROR] src/main/java/org/apache/james/mailbox/store/mail/MailboxMapper.java:[99,5] (extension) RightCurlyAlone: '}' at column 5 should be alone on a line.
```

**Comment by chibenwa** [comment]
???

Build is lost, deployment test

test this please

IMO can be considered :green_apple:

**Comment by chibenwa** [comment]
```
java.lang.IllegalStateException: no known node for /home/jenkins/build/workspace/1205
```

Can be considered :green_apple:

test this please

**Comment by Arsnael** [comment]
CI restart

test this please

**Comment by Arsnael** [comment]
Maybe a rebase? I see there is commits related to ADRs that have been merged since

**Comment by chibenwa** [comment]
## Before

Capture d’écran de 2020-04-10 13-22-11

## After

Capture d’écran de 2020-04-10 13-22-11

## Conclusion

Huge gains: 
 + x4 on mean time for getMessageList
 + x3 on p99 for getMessageList

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**